### PR TITLE
Issue-423 - Render and deploy stan functions' documentation

### LINKED ISSUE #423: Render custom stan functions using Doxygen
**Is your feature request related to a problem? Please describe.**
The current stan documentation (added in #422) follows the suggested format in the stan docs which themselves are based on Doxygen. However, this documentation is not currently rendered so is hard to navigate and in some cases read.

**Describe the solution you'd like**
Using Doxygen to render our documentation similarly to done [here](https://github.com/spinkney/helpful_stan_functions/tree/main). Note there is no official stan support and this example is the only one I have seen. Once this is in place we would also need to clearly link from the main package site.

**Describe alternatives you've considered**
A different documentation system but it is not clear to me what this would be.

@sbfnk @jamesmbaazam might be interested in this for `EpiNow2`.

### PR DESCRIPTION
**Pull request by jamesmbaazam**
## Description

This PR closes #423 by setting up a Doxygen configuration to render the documented stan functions. It also adds a GitHub Actions Workflow to render and deploy the docs to GitHub pages ~but this is currently broken~.

[Describe the changes that you made in this pull request.]

## Checklist

- [x] My PR is based on a package issue and I have explicitly linked it.
- [x] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [x] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [x] I have tested my changes locally.
- [ ] I have added or updated unit tests where necessary.
- [x] I have updated the documentation if required.
- [x] My code follows the established coding standards.
- [x] I have added a news item linked to this PR.
- [ ] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

**Comment by jamesmbaazam** [comment]
TO DOs:
- [ ] Crosscheck github action and make sure it deploys to the right venue
- [ ] Discuss adding doxygen style guidance to CONTRIBUTING.md 
- [ ] Set up custom homepage with markdown to replace index.html??

**Comment by jamesmbaazam** [comment]
Leaving some notes here on installing `doxygen` on MacOS in case someone runs into the issues I ran into:

- `doxygen` installation errors if `qt` is not installed. Fix that with `brew install qt`.
- Building `doxygen` from source has to be preceeded by installation of the GNU `coreutils` package, which makes sure it is installed in the right directory. 
  - `coreutils` is built from source as well. Download it from [here](https://www.gnu.org/software/coreutils/), unzip it, and follow the instructions in the README/INSTALL file in the root directory.
- For `doxygen`, I had to change the minimum MACOS version to 11.00 in the `CMakeLists.txt` file to bypass the installation errors. Search for `MACOS_VERSION_MIN` in `CMakeLists.txt` and change it to `11.00`. See an explanation [here](https://github.com/CefView/QCefView/issues/341#issuecomment-1814708900).
  
NB: When building from source, you may have to use `sudo` in many situations to be able to bypass the permission issues due to I/O processes.

File: .github/workflows/render-stan-docs.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        # Working directory
+        working-directory: "."
+
+    - name: Deploy
+      uses: peaceiris/actions-gh-pages@v3
+      with:
+        github_token: ${{ secrets.GITHUB_TOKEN }}
+        # Default Doxyfile build documentation to html directory.
+        # Change the directory if changes in Doxyfile
+        publish_dir: ./inst/stan/html
```
Comment (jamesmbaazam) [inline .github/workflows/render-stan-docs.yaml:33]: @seabbs You may have to do your magic here to deploy the rendered docs to the desired destination. Wasn't able to do that.

**Comment by jamesmbaazam** [comment]
The current rendered version is nice but could be made easier to navigate by grouping the functions into pages for example.

**Comment by seabbs** [comment]
> Leaving some notes here on installing

We should add these to the developer docs somewhere. Do you have any thoughts as to where?

File: .github/workflows/render-stan-docs.yaml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Based on https://github.com/mattnotmitt/doxygen-action
+name: Doxygen Action
+
+on:
+  pull_request:
```
Comment (seabbs) [inline .github/workflows/render-stan-docs.yaml:5]: We likely only want to deploy on main so may want a different trigger for that (still build on PRs but not try and publish and just push to an environment variable).

File: .github/workflows/render-stan-docs.yaml (hunk 1 above)
Comment (seabbs) [reply .github/workflows/render-stan-docs.yaml:33]: Looking at this it looks like we need to set `destination_dir` and I guess that should be `stan`?

File: .github/workflows/render-stan-docs.yaml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    - name: Doxygen Action
+      uses: mattnotmitt/doxygen-action@v1.9.2
+      with:
+        # Path to Doxyfile
+        doxyfile-path: "./inst/stan/Doxyfile"
+        # Working directory
+        working-directory: "."
+
+    - name: Deploy
+      uses: peaceiris/actions-gh-pages@v3
```
Comment (seabbs) [inline .github/workflows/render-stan-docs.yaml:28]: I think we wannt an if trigger here to check we are on main

File: .github/workflows/render-stan-docs.yaml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    - uses: actions/checkout@v2
+
+    - name: Doxygen Action
+      uses: mattnotmitt/doxygen-action@v1.9.2
+      with:
+        # Path to Doxyfile
+        doxyfile-path: "./inst/stan/Doxyfile"
+        # Working directory
+        working-directory: "."
+
```
Comment (seabbs) [inline .github/workflows/render-stan-docs.yaml:26]: I think we want to save the built docs to an environment variable here if not on main so we can look at PR docs

File: inst/stan/functions/delay_lpmf.stan [hunk 5]
```diff
@@ -81,7 +81,7 @@ real delay_snap_lpmf(array[] int dummy, int start, int end, array[] int obs,
  * @param groups Array of group identifiers.
  * @param start Start index for the range of interest.
  * @param end End index for the range of interest.
- * @commonparams See Common Parameters for `delay_snap_lpmf` and `delay_group_lpmf`
+ * @see \ref common_parameters_page.
```
Comment (seabbs) [inline inst/stan/functions/delay_lpmf.stan:84]: This is good but ideally we would have them inherited if we can (if this is tricky perhaps we need a new issue)

**Review by seabbs — CHANGES_REQUESTED** [review]
This is really great @jamesmbaazam. I am amazed that you didn't need to do more changes to the stan docs to get this working. I made a few suggestions for the Action that should help us look at the output etc.

File: inst/stan/Doxyfile [hunk 6]
```diff
@@ -0,0 +1,2927 @@
...
+# pixels and the maximum width should not exceed 200 pixels. Doxygen will copy
+# the logo to the output directory.
+
+PROJECT_LOGO           =
+
+# With the PROJECT_ICON tag one can specify an icon that is included in the tabs
+# when the HTML document is shown. Doxygen will copy the logo to the output
+# directory.
+
+PROJECT_ICON           = /Users/lshja16/Documents/Git_repositories/_Work/LSHTM/Epinowcast/epinowcast/man/figures/logo.png
```
Comment (seabbs) [inline inst/stan/Doxyfile:70]: if we want to build this on actions I would expect this needs to be relative link

File: inst/stan/functions/delay_lpmf.stan (hunk 5 above)
Comment (jamesmbaazam) [reply inst/stan/functions/delay_lpmf.stan:84]: True. This was a lazy first pass but not ideal at all for navigation. I think we might be able to do it with [`copydoc`](https://www.doxygen.nl/manual/commands.html#cmdcopydoc). I'll look into it.

**Comment by jamesmbaazam** [comment]
> We should add these to the developer docs somewhere. Do you have any thoughts as to where?

Yes, the plan was to move it somewhere more visible. Do we have a wiki? We could post it there and link to it from the CONTRIBUTING.md

File: inst/stan/Doxyfile (hunk 6 above)
Comment (jamesmbaazam) [reply inst/stan/Doxyfile]: Oopsie. I missed this one. I changed all the others to a relative link. This is a remnant of me testing it locally. I've changed it in 41b01bc82a93284c2c55526f78a598d2b19381f0.

**Comment by seabbs** [comment]
@jamesmbaazam will circle back but we need to be careful: https://github.com/epinowcast/epinowcast/tree/gh-pages

This PR has killed the overall site!

**Comment by seabbs** [comment]
Right no worries I reverted the push to `gh-pages`. Easily done. Not sure what kind of protection we can put on that.. hm maybe make it so only the bot can push to `gh-pages`?

**Comment by seabbs** [comment]
> Yes, the plan was to move it somewhere more visible. Do we have a wiki? We could post it there and link to it from the CONTRIBUTING.md

We don't. This could also go in the community wide `.github` advice as well I guess

**Comment by jamesmbaazam** [comment]
> Right no worries I reverted the push to `gh-pages`. Easily done. Not sure what kind of protection we can put on that.. hm maybe make it so only the bot can push to `gh-pages`?

Sorry about that. Yeah, some branch protection rules to prevent his would be helpful moving forward.

**Comment by seabbs** [comment]
@jamesmbaazam I made the suggested GH action changes in #502.

File: .github/workflows/render-stan-docs.yaml [hunk 7]
```diff
@@ -0,0 +1,49 @@
+# Based on https://github.com/mattnotmitt/doxygen-action
+name: Doxygen Action
```
Comment (seabbs) [inline .github/workflows/render-stan-docs.yaml:2]:
```suggestion
name: Render Stan documentation
```

File: inst/stan/functions/common_params.stan [hunk 8]
```diff
@@ -0,0 +1,56 @@
+/**
+ * \page common_parameters_page Common Parameters
```
Comment (seabbs) [inline inst/stan/functions/common_params.stan:2]: Can we rename this page to make it clear what they are common parameters for? i.e delay_parameter-docs or something?

**Review by seabbs — COMMENTED** [review]
Very nearly there I think! Shall we make a custom index and linking to the site etc part of a new issue (ie prettify stan docs) vs doing it here?

**Comment by seabbs** [comment]
Also before merging this James can you bump yourself up to an author in the DESCRIPTION?

File: inst/stan/functions/common_params.stan (hunk 8 above)
Comment (jamesmbaazam) [reply inst/stan/functions/common_params.stan]: Good point. I've renamed the page here b95533e047bfc3611ce37bac82cecfbca7555047.

**Comment by jamesmbaazam** [comment]
### Issues with doxygen's infrastructure for documentation reuse

TLDR: doxygen's `@copydoc` allows you to use shared parameter documentation but is not as smart as the Roxygen2 `@inheritparams` to order parameter names according to the function signature.

#### Issues encountered

- If you want the arguments in the order that they appear in the function signature, you end up with two "Parameter" headings, one coming from the documentation of specific params using `@param` and another coming from `@copydoc`. Unfortunately, there is no way to configure `@copydoc` to at least drop its section title.
- You can get one heading if you document the other parameters after the `@copydoc` call but that means the parameters are not in the order in which they appear in the function signature. This is the option currently implemented here.

File: inst/stan/functions/delay_lpmf.stan [hunk 9]
```diff
@@ -77,14 +72,37 @@ real delay_snap_lpmf(array[] int dummy, int start, int end, array[] int obs,
...
+ * @copydoc common_parameters_delay_lpmf_funcs
+ *
  * @param groups Array of group identifiers.
  * @param start Start index for the range of interest.
- * @param end End index for the range of interest.
- * @commonparams See Common Parameters for `delay_snap_lpmf` and `delay_group_lpmf`
- * 
+ * @copydoc common_parameters_page
+ *
+ * @param t
```
Comment (jamesmbaazam) [inline inst/stan/functions/delay_lpmf.stan:82]: @seabbs Could you document `t` and check that the others I added are accurate?

File: .github/workflows/render-stan-docs.yaml [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+    # Steps represent a sequence of tasks that will be executed as part of the job
+    steps:
+    # Checks-out your repository under $GITHUB_WORKSPACE, so your job can access it
+    - uses: actions/checkout@v2
+
+    - name: Doxygen Action
+      uses: mattnotmitt/doxygen-action@v1.9.2
+      with:
+        # Path to Doxyfile
```
Comment (seabbs) [inline .github/workflows/render-stan-docs.yaml:28]: In the output for this I see it thinks it can't access `functions` and that no html is found at `inst/stan/html`. Thoughts @jamesmbaazam?

**Review by seabbs — COMMENTED** [review]
> ### Issues with doxygen's infrastructure for documentation reuse

TLDR: doxygen's `@copydoc` allows you to use shared parameter documentation but is not as smart as the Roxygen2 `@inheritparams` to order parameter names according to the function signature.

#### Issues encountered

- If you want the arguments in the order that they appear in the function signature, you end up with two "Parameter" headings, one coming from the documentation of specific params using `@param` and another coming from `@copydoc`. Unfortunately, there is no way to configure `@copydoc` to at least drop its section title.
- You can get one heading if you document the other parameters after the `@copydoc` call but that means the parameters are not in the order in which they appear in the function signature. This is the option currently implemented here. 

A bit frustating but seems reasonable as a solution. 

I have added docs for `t` and reviewed the other newly added docs.

File: .github/workflows/render-stan-docs.yaml (hunk 2 above)
Comment (jamesmbaazam) [reply .github/workflows/render-stan-docs.yaml:6]: Resolved through #502 by @seabbs.

File: .github/workflows/render-stan-docs.yaml (hunk 4 above)
Comment (jamesmbaazam) [reply .github/workflows/render-stan-docs.yaml:32]: Resolved through #502

File: .github/workflows/render-stan-docs.yaml (hunk 10 above)
Comment (jamesmbaazam) [reply .github/workflows/render-stan-docs.yaml:28]: Going try to debug locally with [act](https://github.com/nektos/act). There may be a weird interaction between the paths specified in the `Doxyfile` and the paths specified here. I'll check now.

Comment (seabbs) [reply .github/workflows/render-stan-docs.yaml:28]: > This has been a weirdly fun exercise. I really appreciate {Roxygen2} now 😆.

Haha I know right. Try Julia Docstrings (or Python) so painful!

> I think the R stan community n

Agree they just need a better focus on user packaging across the board IMO. 

> I could explore it later. Or have you?

I haven't looked at this and that would be great if we could get it to work

Comment (jamesmbaazam) [reply .github/workflows/render-stan-docs.yaml:28]: > In the output for this I see it thinks it can't access `functions` and that no html is found at `inst/stan/html`. Thoughts @jamesmbaazam?

I think this is fixed now as of 29441a2e5e83b2e02ed443d83bcde8b8752b4492.

**Review by seabbs — APPROVED** [review]
This is great. Functionally it is good to go. We just need a news item update for this PR.

**Comment by jamesmbaazam** [comment]
> This is great. Functionally it is good to go. We just need a news item update for this PR.

Done!

**Review by seabbs — APPROVED** [review]
